client/cli_crypto_service.go

```diff
@@ -10,62 +11,77 @@ import (
-func NewCryptoService(gun string, keyStore *trustmanager.KeyFileStore) *CryptoService {
-	return &CryptoService{gun: gun, keyStore: keyStore}
+// RSACryptoService implements Sign and Create, holding a specific GUN and keystore to
+// operate on
+type RSACryptoService struct {
+	genericCryptoService
+}
+
+// ECDSACryptoService implements Sign and Create, holding a specific GUN and keystore to
+// operate on
+type ECDSACryptoService struct {
+	genericCryptoService
+}
+
+// NewRSACryptoService returns an instance of CryptoService
+func NewRSACryptoService(gun string, keyStore *trustmanager.KeyFileStore, passphrase string) *RSACryptoService {
+	return &RSACryptoService{genericCryptoService{gun: gun, keyStore: keyStore, passphrase: passphrase}}
 }
 
 // Create is used to generate keys for targets, snapshots and timestamps
-func (ccs *CryptoService) Create(role string) (*data.PublicKey, error) {
+func (ccs *RSACryptoService) Create(role string) (*data.PublicKey, error) {
 	privKey, err := trustmanager.GenerateRSAKey(rand.Reader, rsaKeySize)
 	if err != nil {
 		return nil, fmt.Errorf("failed to generate RSA key: %v", err)
 	}
 
 	// Store the private key into our keystore with the name being: /GUN/ID.key
-	ccs.keyStore.AddKey(filepath.Join(ccs.gun, privKey.ID()), privKey)
+	err = ccs.keyStore.AddKey(filepath.Join(ccs.gun, privKey.ID()), privKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add key to filestore: %v", err)
+	}
 
-	return data.PublicKeyFromPrivate(*privKey), nil
-}
+	logrus.Debugf("generated new RSA key for role: %s and keyID: %s", role, privKey.ID())
 
-// SetPassphrase tells the cryptoservice the passphrase. Use only if the key needs
-// to be decrypted.
-func (ccs *CryptoService) SetPassphrase(passphrase string) {
-	ccs.passphrase = passphrase
+	return data.PublicKeyFromPrivate(*privKey), nil
 }
 
-// Sign returns the signatures for data with the given root Key ID, falling back
-// if not rootKeyID is found
-func (ccs *CryptoService) Sign(keyIDs []string, payload []byte) ([]data.Signature, error) {
+// Sign returns the signatures for the payload with a set of keyIDs. It ignores
+// errors to sign and expects the called to validate if the number of returned
+// signatures is adequate.
+func (ccs *RSACryptoService) Sign(keyIDs []string, payload []byte) ([]data.Signature, error) {
 	// Create hasher and hash data
 	hash := crypto.SHA256
 	hashed := sha256.Sum256(payload)
 
 	signatures := make([]data.Signature, 0, len(keyIDs))
-	for _, fingerprint := range keyIDs {
+	for _, keyid := range keyIDs {
 		// ccs.gun will be empty if this is the root key
-		keyName := filepath.Join(ccs.gun, fingerprint)
+		keyName := filepath.Join(ccs.gun, keyid)
 
 		var privKey *data.PrivateKey
 		var err error
 		var method string
 
-		// Read PrivateKey from file
+		// Read PrivateKey from file.
+		// TODO(diogo): This assumes both that only root keys are encrypted and
+		// that encrypted keys are always X509 encoded
```

Comment:
I do think we'll want to change this eventually - I can imagine use cases where the root keys will not have a passphrase (i.e. non-interactive use). I think it would be a good idea to add a flag to the `genericCryptoService` structure to use instead of using the passphrase for deciding whether it's a root key or not.